Storage/src/Connection/Rest.php

```diff
@@ -599,4 +640,96 @@ protected function supportsBuiltinCrc32c()
+    {
+        $retryMap = [
+            'projects.resources.serviceAccount' => 'serviceaccount',
+            'projects.resources.hmacKeys' => 'hmacKey',
+            'bucketAccessControls' => 'bucket_acl',
+            'defaultObjectAccessControls' => 'default_object_acl',
+            'objectAccessControls' => 'object_acl'
+        ];
+        $retryResource = isset($retryMap[$resource]) ? $retryMap[$resource] : $resource;
+        $options['restRetryFunction'] = $this->restRetryFunction ?? $this->getRestRetryFunction(
+            $retryResource,
+            $method,
+            $options
+        );
+
+        $options = $this->addRetryHeaderLogic($options);
+
+        return $this->traitSend($resource, $method, $options);
+    }
+
+    /**
+     * Adds the retry headers to $args which amends retry hash and attempt
+     * count to the required header.
+     * @param array $args
+     * @return array
+     */
+    private function addRetryHeaderLogic(array $args)
+    {
+        $invocationId = Uuid::uuid4()->toString();
+        $args['retryHeaders'] = self::getRetryHeaders($invocationId, 1);
+
+        $currentListener = $args['restRetryListener'] ?? null;
+
+        // Adding callback logic to update headers while retrying
+        $args['restRetryListener'] = function (
+            \Exception $e,
+            $retryAttempt,
+            $arguments
+        ) use (
+            $invocationId,
+            $currentListener
+        ) {
+            $changes = self::getRetryHeaders($invocationId, $retryAttempt + 1);
+            $request = $arguments[0];
+            $headerLine = $request->getHeaderLine(AgentHeader::AGENT_HEADER_KEY);
+
+            // An associative array to contain final header values as
+            // $headerValueKey => $headerValue
+            $headerElements = [];
+
+            // Adding existing values
+            $headerLineValues = explode(' ', $headerLine);
+            foreach ($headerLineValues as $value) {
+                $key = explode('/', $value)[0];
+                $headerElements[$key] = $value;
+            }
+
+            // Adding changes with replacing value if $key already present
+            foreach ($changes as $change) {
+                $key = explode('/', $change)[0];
+                $headerElements[$key] = $change;
+            }
+            $arguments[0] = Utils::modifyRequest($request, [
+                'set_headers' => [
+                    AgentHeader::AGENT_HEADER_KEY => implode(' ', $headerElements)
+                ]
+            ]);
+
+            // In some cases there might be a listener present
+            // So, we want to execute that after incrementing the attempt count
+            // Ex in case of downloads
+            if (!is_null($currentListener)) {
+                $arguments = call_user_func_array($currentListener, [
+                    $e, $retryAttempt, $arguments
+                ]);
+            }
```

Comment:
It isn't really necessary for us to allow for a `restRetryListener` option to be supplied by our customer. Unless there's a known use-case, it would be better to only use this option internally. See https://github.com/googleapis/google-cloud-php/pull/6059

Reply:
So, this is done to support downloads.

In this case, the downloads need a retryListener to support the resumable nature(Range headers).

We also need the RetryListener in Rest.php so that we can inject the request hash and the attempt count headers.

I can think of another way where we abstract the logic in Rest.php and use it inside the listener of the downloads. Let me know if that sounds cleaner?

Reply:
Just to be clear, I am talking about the `$currentListener` variable, not the user passing it in the constructor.

Reply:
But I am in favour of keeping this `@internal`.

Reply:
I've added a PR which shows how we can cleanly avoid including special logic to call nested listener options 

https://github.com/googleapis/google-cloud-php/pull/6073

Reply:
Merged #6073 Thanks for the PR.